### DISCUSSION TITLE
Design Equations

### DISCUSSION THREAD

**Discussion by sboris369** [Eomys/pyleecan#390] [Q&A]
From what source(s) are the design equations used in Pyleecan derived from?.

**Comment by BonneelP** [comment]
Hello,

Thank you for your interest in pyleecan. Can you be more specific : which design equations you are interested in ? The point of pyleecan is to have several model to compute the different physics and options to take into account (or not) some phenomena.

Best regards,
Pierre

**Reply by sboris369** [reply]
Hi Pierre
source(s)  refers to books from certain authors or even published papers.
I have numerous books on machine design and  each author uses their own
set of symbols in their design equations. If pyleecan uses it's own list of symbols
I would like to adopt those, so that anything I write  conforms to that set.
This way anyone reading or modifying  the code will be sure  of what parameter is being computed.

For what I want to do in pyleecan, is to add a wound rotor synchronous generator to the list
of machine types deriving it from MachineWRSM or MachineSync. Knowing from what sources the
equations came from would help in reusing everything that's already written.

Thanks for your help.
Steve

**Reply by BonneelP** [reply]
Hello Steve,

The first place you can look for information on the quantities/symbols we are using is this folder:
https://github.com/Eomys/pyleecan/tree/master/pyleecan/Generator/ClassesRef
In pyleecan all the class files are automatically generated from csv files stored in this folder. Each input and output of pyleecan is explained with units, type, description (min and max sometimes). So In particular https://github.com/Eomys/pyleecan/tree/master/pyleecan/Generator/ClassesRef/Output details all the Outputs.
![image](https://user-images.githubusercontent.com/20533034/119896213-043ec700-bf3f-11eb-98c2-9594b857208a.png)

The second place you can look for reference is the code itself. In some test and method we sometimes includes the model reference for instance:
https://github.com/Eomys/pyleecan/blob/master/pyleecan/Methods/Simulation/EEC_PMSM/solve_EEC.py
https://github.com/Eomys/pyleecan/blob/master/Tests/Validation/Magnetics/test_torque.py
Most of the functions/methods also have doc-string explaining the expected input / output with type, unit and description:
https://github.com/Eomys/pyleecan/blob/master/pyleecan/Methods/Machine/LamSlotWind/comp_resistance_wind.py
We also have few information on the website but not as many as I would like: https://pyleecan.org/coding.convention.html

If you find some equation without reference feel free to ask and we will see if we can find the source back. 

Great to know that you are interested in working on WRSM generator. You should be able to define this kind of machine with the GUI but we lack validation cases for this kind of machines. So if you have any interesting reference publication / validation data to share on this topology it would be really helpful. 

I hope things are more clear now, if you need further information feel free to ask.
Best regards,
Pierre

**Reply by sboris369** [reply]
Hi Pierre

Thanks for your reply.
ClassRef is the first place I started looking and will mine everything from there that I find.
I like the fact that all the details are stored in csv files, make everything so much easier.

I have been going over the coding conventions so anything I write complies with Pep8.
Could you suggest any guidelines on how much documentation to put in the code itself and
how much to put in external files. I tend to put a lot of info in the code itself.

As I go thru the process of adding this machine type,  I will be documenting all the steps that are needed.
I will submit it to the doc-repo when ready, so it's available to anyone that's interested in adding a new machine type.

Thanks again,
Steve

**Reply by BonneelP** [reply]
Hello,

The main question is "for whom I am writing documentation for ?". Two kinds of people in fact: the one that will use your code without know how it works and the one who will try to challenge/improve your code. So I would say:
- In a method like "comp_something" the reference from where the equation came from (with page or equation number) + related assumptions and units
- In pyleecan-doc (aka the website pyleecan.org) the general conventions like what does it imply to have a "generator" instead of a motor ? For how to add a new machine type, you can take inspiration from: https://pyleecan.org/tuto.add.slot.html (if needed I can help for the GUI part)
- In Tests/Validation a full simulation with a comparison with experimental data or publication with the proper link and step by step check (the Tests are run regularly to make sure that your code is still compatible with pyleecan evolution, so we need to know from where the results came from if we manage to remove/improve a constraint/assumptions one day). It will also be the comparison reference if different models are added for the same purpose.
- A jupyter notebook tutorial with full examples and equation in latex and all the references gather in a single place. The tutorial will be added to the [website](https://pyleecan.org/tutorials.html) and can be the entry point for your work for people that want to use your models (more scientifics than developers).

I would say this repartition is the best case scenario, but take your time to construct step by step your functions. You can even create several PR so we can discuss along your development and help you find the proper documentation/implementation. 

Best regards,
Pierre
